chalk-ir/src/visit.rs

```diff
@@ -0,0 +1,404 @@
+        outer_binder: DebruijnIndex,
+    ) -> Self::Result {
+        lifetime.super_visit_with(self.as_dyn(), outer_binder)
+    }
+
+    /// Invoked for every program clause. By default, recursively visits the goals contents.
+    fn visit_program_clause(
+        &mut self,
+        clause: &ProgramClause<I>,
+        outer_binder: DebruijnIndex,
+    ) -> Self::Result {
+        clause.super_visit_with(self.as_dyn(), outer_binder)
+    }
+
+    /// Invoked for every goal. By default, recursively visits the goals contents.
+    fn visit_goal(&mut self, goal: &Goal<I>, outer_binder: DebruijnIndex) -> Self::Result {
+        goal.super_visit_with(self.as_dyn(), outer_binder)
+    }
+
+    /// If overridden to return true, then visiting will panic if a
+    /// free variable is encountered. This should be done if free
+    /// type/lifetime variables are not expected.
+    fn forbid_free_vars(&self) -> bool {
+        false
+    }
+
+    /// Invoked for `TyData::BoundVar` instances that are not bound
+    /// within the type being visited over:
+    fn visit_free_var_ty(
+        &mut self,
+        bound_var: BoundVar,
+        outer_binder: DebruijnIndex,
+    ) -> Self::Result {
+        if self.forbid_free_vars() {
+            panic!(
+                "unexpected free variable `{:?}` with outer binder {:?}",
+                bound_var, outer_binder
+            )
+        } else {
+            Self::Result::new()
+        }
+    }
+
+    fn visit_bound_var_ty(
+        &mut self,
+        _bound_var: BoundVar,
+        _outer_binder: DebruijnIndex,
+    ) -> Self::Result {
+        Self::Result::new()
+    }
+
+    /// As `visit_free_var_ty`, but for lifetimes.
+    fn visit_free_var_lifetime(
+        &mut self,
+        bound_var: BoundVar,
+        outer_binder: DebruijnIndex,
+    ) -> Self::Result {
+        if self.forbid_free_vars() {
+            panic!(
+                "unexpected free variable `{:?}` with outer binder {:?}",
+                bound_var, outer_binder
+            )
+        } else {
+            Self::Result::new()
+        }
+    }
+
+    fn visit_bound_var_lifetime(
+        &mut self,
+        _bound_var: BoundVar,
+        _outer_binder: DebruijnIndex,
+    ) -> Self::Result {
+        Self::Result::new()
+    }
+
+    /// If overriden to return true, we will panic when a free
+    /// placeholder type/lifetime is encountered.
+    fn forbid_free_placeholders(&self) -> bool {
+        false
+    }
+
+    /// Invoked for each occurrence of a placeholder type; these are
+    /// used when we instantiate binders universally.
+    fn visit_free_placeholder_ty(
+        &mut self,
+        universe: PlaceholderIndex,
+        _outer_binder: DebruijnIndex,
+    ) -> Self::Result {
+        if self.forbid_free_placeholders() {
+            panic!("unexpected placeholder type `{:?}`", universe)
+        } else {
+            Self::Result::new()
+        }
+    }
+
+    fn visit_alias_eq(
```

Comment:
Hmm, do we really want this accessor? I think we don't have similar one for other kinds of specific goals